runtime/etc/src/mli/common_code.c

```diff
@@ -228,23 +211,117 @@ void chpl_mli_close(void* socket) {
+
+  while (-1 == result && errno == EAGAIN) {
+    chpl_mli_debugf("Socket busy, retry: %d\n", retry);
+    result = chpl_mli_zmq_wire(socket, buffer, bytes, push);
+    if (retry++ > 5) {
+      chpl_mli_debugf("%s\n", "Too many retries!");
+      chpl_mli_terminate();
+    }
+  }
+
+  // TODO: We need to handle socket errors more robustly.
+  if (-1 == result) {
+    chpl_mli_debugf("Failed to push to socket: %p\n", socket);
+    chpl_mli_debugf("%s\n", zmq_strerror(errno));
+    chpl_mli_terminate();
+  }
+
+  return result;
+}
+
+static
+int chpl_mli_push(void* socket, void* buffer, size_t bytes) {
+  chpl_mli_debugf("%zu bytes at %p to %p\n",
+                  bytes,
+                  buffer,
+                  socket);
+
+  return chpl_mli_zmq_move(socket, buffer, bytes, 1);
+}
+
+static
+int chpl_mli_pull(void* socket, void* buffer, size_t bytes) {
+  chpl_mli_debugf("%zu bytes at %p to %p\n",
+                  bytes,
+                  buffer,
+                  socket);
+
+  return chpl_mli_zmq_move(socket, buffer, bytes, 0);
+}
+
+static
+int chpl_mli_push_ack(void* socket) {
+  return chpl_mli_push(socket, NULL, 0);
 }
 
 static
-int chpl_mli_pull(void* socket, void* buffer, size_t bytes, int flags) {
-  return zmq_recv(socket, buffer, bytes, flags);
+int chpl_mli_pull_ack(void* socket) {
+  return chpl_mli_pull(socket, NULL, 0);
+}
+
+static
+int chpl_mli_push_byte_buffer(void* socket, chpl_byte_buffer* obj) {
+  chpl_mli_push(socket, &obj->size, sizeof(obj->size));
+  chpl_mli_pull_ack(socket);
+  chpl_mli_debugf("Byte buffer length: %" PRIu64 "\n", obj->size);
+
+  chpl_mli_push(socket, obj->data, obj->size);
+  chpl_mli_pull_ack(socket);
+
+  // TODO: User data may not be trustable? E.g. not null terminated...
+  // TODO: Truncate beyond a certain length.
+  chpl_mli_debugf("Byte buffer data: %s\n", obj->data);
+
+  return 0;
+}
+
+static
+int chpl_mli_pull_byte_buffer(void* socket, chpl_byte_buffer* obj) {
+  chpl_mli_pull(socket, &obj->size, sizeof(obj->size));
+  chpl_mli_push_ack(socket);
+  chpl_mli_debugf("Byte buffer length: %" PRIu64 "\n", obj->size);
+
+  obj->data = chpl_mli_malloc(obj->size + 1);
+  obj->isOwned = true;
+
+  // TODO: Need to handle synchronous termination.
+  if (NULL == obj->data) {
+    chpl_mli_debugf("%s\n", "Failed to allocate buffer!");
+    chpl_mli_terminate();
+  }
+
+  chpl_mli_pull(socket, obj->data, obj->size);
+  chpl_mli_push_ack(socket);
+
+  obj->data[obj->size] = 0;
+
+  chpl_mli_debugf("Byte buffer data: %s\n", obj->data);
+
+  return 0;
 }
 
 // Determine connection information for that socket.
 static
 char * chpl_mli_connection_info(void* socket) {
   // Determine port used by ZMQ for socket.
   size_t lenPort = 256;
-  char* portRes = (char *)mli_malloc(lenPort);
+  char* portRes = (char *)chpl_mli_malloc(lenPort);
   int portErr = zmq_getsockopt(socket, ZMQ_LAST_ENDPOINT,
                                portRes, &lenPort);
+  (void) portErr;
```

Comment:
What does this do?

Reply:
Silences compiler warnings about it not being used.

Reply:
I think the normal pattern for this is just:

```c
(void)zmq_getsockopt(socket, ZMQ_LAST_ENDPOINT, portRes, &lenPort);
```

Reply:
Added a comment above `(void) portErr;` instead.

Reply:
It seems like we should either check the value returned or not store it, right?